Migrate url scheme

ayyazdaniaryan opened the pull request.

vincentfretin commented on assembl/static2/js/app/components/login/socialMediaLogin.jsx:
```diff
@@ -0,0 +1,41 @@
+            return (<li key={provider.name}>
+              <form id={provider.name} method="get" action={provider.login} >
+                {Object.keys(provider.extra).map((k) => {
+                  return (<input key={provider.name + k} type="hidden" name={k} value={provider.extra[k]} />);
```
You can just use key={k} here, the key only need to be unique locally, not globally.

paolina92 replied:
haha I did the same error ;) Now we know.

ayyazdaniaryan replied:
Yes, did that initially. Sadly, the keys in the `provider.extra` is not unique, especially in a SAML case. There can be 2 different SAML providers names with the same keys. So, appended the `provider.name` to maintain uniqueness locally.

vincentfretin commented on assembl/static2/js/app/pages/changePassword.jsx:
```diff
@@ -1,22 +1,90 @@
+    // console.log('handle password', e);
+    const s = {};
+    s[e.target.name] = e.target.value;
+    this.setState(s);
```
You really can do this in one line:
```
this.setState({ [e.target.name]: e.target.value });
```

ayyazdaniaryan replied:
Oops, forgot to update this one. Thanks for pointing it out.

vincentfretin commented on assembl/static2/js/app/pages/signup.jsx:
```diff
@@ -1,22 +1,37 @@
+
   render() {
-    const isAccountCreated = location.hash === '#success';
+    // console.log(this.props);
```
please remove commented code

vincentfretin commented on assembl/static2/js/app/pages/signup.jsx:
```diff
@@ -1,22 +1,37 @@
-            {!isAccountCreated && <SignupForm />}
-            {isAccountCreated && <SignUpConfirm />}
+            {!signupConfirm && <SignupForm />}
+            {signupConfirm && <SignUpConfirm />}
```
When there is the two cases, I like better the following syntax:
```
{signupConfirm ? <SignUpConfirm /> : <SignupForm />}
```

ayyazdaniaryan replied:
Haha, fair enough. I've made a fair bit of comments on #90 to use this style as well, since it appears both you and @paolina92 like this approach more.

vincentfretin commented on assembl/static2/js/app/utils/errors.js:
```diff
@@ -0,0 +1,6 @@
+/*
+  All special erros are placed here. Creating special errors is very useful
```
typo: erros => errors

vincentfretin commented on assembl/static2/js/app/utils/globalFunctions.js:
```diff
@@ -48,4 +66,12 @@ export const getNumberOfDays = (date1, date2) => {
+/*
+  Handrolled instead of using lowdash
+  Because lowdash/capitalize lowercases everything else
+*/
```
lodash, not lowdash :)

ayyazdaniaryan replied:
Me no english 😂

vincentfretin commented on assembl/tests/views_tests/test_routes.py:
```diff
@@ -0,0 +1,296 @@
+@pytest.mark.xfail
+def test_route_discussion_home(discussion, test_app):
+    """/slug/home"""
+    # TODO: Implement this route!
```
This is still not implemented?

maparent replied:
It is, I suspect this is an unfinished test.

ayyazdaniaryan replied:
It is. Stale comment.

vincentfretin commented on assembl/views/__init__.py:
```diff
@@ -226,10 +282,25 @@ def process_export_list(ls):
+    if messages:
+        kwargs['messages'] = '<br />'.join(messages)
+
+    admin_email = config.get('assembl.admin_email') or "assembl.admin@bluenove.com"
```
You really want a hard coded value here?

maparent replied:
It's a fallback to the config file. The alternative is to refuse to start assembl, saying that it is not properly configured... which may be a good idea, actually. Other opinions?

ayyazdaniaryan replied:
I think it's overkill to not start Assembl if the admin_email isn't available. But, I'm not clear on downstream implications. I am okay with an exception raised here if admin email doesn't exist.

vincentfretin commented on assembl/views/__init__.py:
```diff
@@ -253,7 +324,8 @@ def process_export_list(ls):
         activate_tour=str(config.get('activate_tour') or False).lower(),
         providers=json.dumps(providers),
-        translations=codecs.open(jedfilename, encoding='utf-8').read()
+        translations=codecs.open(jedfilename, encoding='utf-8').read(),
```
FYI, codecs.open is quite old, in Python 2.7, you can now import open from the io package (backported from python 3.) 
```
from io import open
```
and it supports the encoding param. The builtin open function doesn't support encoding indeed.

maparent replied:
Thanks for pointing this out! Not sure if it belongs in this PR though?

vincentfretin replied:
You can keep the line as it is, this is still a valid code.

vincentfretin commented on assembl/views/discussion/views.py:
```diff
@@ -182,16 +182,71 @@ def react_view(request):
     """
     old_context = base_default_context(request)
-    
+    user_id = authenticated_userid(request) or Everyone
```
you should start using request.authenticated_userid
The authenticated_userid function is deprecated.

maparent replied:
Same comment: That's best done in a single commit, outside this PR. Good catch otherwise!

vincentfretin commented on doc/testing.rst:
```diff
@@ -257,6 +257,18 @@ Assembl has several core fixtures that are important to note, in order to run th
+
+To get started with testing in Assembl, please refer to the :ref:`TestingAnchor`.
+
+
```
Why there are two Getting Started here?

ayyazdaniaryan replied:
Aaah. Must be from the rebase that I did. I'll fix.

paolina92 commented on assembl/static2/js/app/components/common/avatar.jsx:
```diff
@@ -2,34 +2,35 @@ import React from 'react';
     const { debateData } = this.props.debate;
-    const { users, usersLoading, usersError } = this.props.users;
-    const { rootPath, connectedUserId } = this.props.context;
+    const slug = { slug: debateData.slug };
```
Use `debateData.slug` line 17 and remove this one.

paolina92 commented on assembl/static2/js/app/components/common/avatar.jsx:
```diff
@@ -2,34 +2,35 @@ import React from 'react';
       <div className="right avatar">
         {!connectedUserId &&
-          <Link to={`${rootPath}${debateData.slug}/login`}>
+          <Link to={`${Routes.getContextual('login', slug)}?next=${Routes.get('home', slug)}`}>
```
After the merge, we should change links every where in the application :-(
`Routes.get('home', { slug: debateData.slug })`

ayyazdaniaryan replied:
I've done that in most places on the new front-end. I might have missed a few, but I've already made most of the migration.

paolina92 commented on assembl/static2/js/app/components/login/assemblLogin.jsx:
```diff
@@ -3,38 +3,42 @@ import { connect } from 'react-redux';
 import { Link } from 'react-router';
 import { form, FormGroup, FormControl, Button } from 'react-bootstrap';
 import { Translate, I18n } from 'react-redux-i18n';
+import { Routes } from '../../routes';
```
Same comment than before.

paolina92 commented on assembl/static2/js/app/components/login/sendPwdForm.jsx:
```diff
@@ -2,27 +2,64 @@ import React from 'react';
+
+  handleInput(e) {
+    inputHandler(this, e);
+    const el = document.getElementById('sendPwdForm-error');
```
Can you use `ref` instead of `getElementById`
https://facebook.github.io/react/docs/refs-and-the-dom.html

ayyazdaniaryan replied:
I'm not sure still how to use ref. My questions regarding ref on #90 was not answered. So, please, can you give me a simpler explanation on how ref is used? Thanks.

vincentfretin replied:
I think it's well explained here https://facebook.github.io/react/docs/refs-and-the-dom.html

paolina92 commented on assembl/static2/js/app/components/login/sendPwdForm.jsx:
```diff
@@ -2,27 +2,64 @@ import React from 'react';
+  handleInput(e) {
+    inputHandler(this, e);
+    const el = document.getElementById('sendPwdForm-error');
+    if (el) { el.innerHTML = ''; }  // BUG: Only shows once; if fails beyond first time, no error shown!
```
Can you use `dangerouslySetInnerHTML` instead of `innerHTML `
https://facebook.github.io/react/docs/dom-elements.html

We should fix the bug instead of commit `//BUG...`

ayyazdaniaryan replied:
This code all dies when I migrate to the alert system. So, I kept a note for myself, but in the end, not so important.

paolina92 commented on assembl/static2/js/app/components/login/sendPwdForm.jsx:
```diff
@@ -2,27 +2,64 @@ import React from 'react';
         <div className="box">
-          <form className="resendPwd" method="POST" action={`${rootPath}${debateData.slug}/changePassword#success`}>
+          <form className="resendPwd" onSubmit={this.submitHandler}>
+            <input type="hidden" name="referer" value="v2" />
```
Why `value="v2"`?

ayyazdaniaryan replied:
This was the initial implementation that Quentin started. I've kept it for compatibility, but the backend (I don't think) makes further checks for this. I'll double check, if it doesn't, I'll remove this, otherwise, it is required to have in order to land on the /home page after a login.

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
 import { Translate, I18n } from 'react-redux-i18n';
 import { form, FormGroup, FormControl, Button } from 'react-bootstrap';
 import { Link } from 'react-router';
+import { Routes } from '../../routes';
```
Same comment than before

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
 class SignupForm extends React.Component {
+  static errorMessageHandler(data) {
+    if (data.success === false) {
+      // Add more cases to this switch statement, as needed
```
Remove this line no?

paolina92 commented on assembl/static2/js/app/components/login/sendPwdForm.jsx:
```diff
@@ -2,27 +2,64 @@ import React from 'react';
+              <FormControl
+                type="text" name="identifier" required
+                placeholder={I18n.t('login.username')} onChange={this.handleInput}
+              />
```
Each attribute on 1 line or all attributes on 1 line

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
-              <FormControl type="text" name="name" required placeholder={I18n.t('login.fullName')} />
+              <FormControl
+                type="text" name="name" required
+                placeholder={I18n.t('login.fullName')} onChange={this.handleInput}
```
Each attribute on 1 line or all attributes on 1 line

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
+            <FormGroup>
+              <FormControl
+                type="text" name="username"
+                placeholder={I18n.t('login.userName')} onChange={this.handleInput}
```
Same here

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
-              <FormControl type="email" name="email" required placeholder={I18n.t('login.email')} />
+              <FormControl
+                type="email" name="email" required
+                placeholder={I18n.t('login.email')} onChange={this.handleInput}
```
and here

paolina92 commented on assembl/static2/js/app/components/login/signupForm.jsx:
```diff
@@ -3,41 +3,109 @@ import { connect } from 'react-redux';
+              <FormControl
+                type="password" name="password2" required
+                placeholder={I18n.t('login.password2')} onChange={this.handleInput}
+              />
```
same here

paolina92 commented on assembl/static2/js/app/components/login/socialMediaLogin.jsx:
```diff
@@ -0,0 +1,41 @@
+//     )
+//   }
+// }
+
```
Remove commented code

paolina92 commented on assembl/static2/js/app/components/login/socialMediaLogin.jsx:
```diff
@@ -0,0 +1,41 @@
+
+export class SocialMedia extends React.Component {
+  // TODO: className="`${provider.type}-bg`" on button
+  render() {
```
Remove TODO before the commit and do it if possible

ayyazdaniaryan replied:
Okay, should have been more clear. TODO: Add CSS to social logins (pair programming with pauline)

paolina92 commented on assembl/static2/js/app/pages/changePassword.jsx:
```diff
@@ -1,22 +1,90 @@
+  }
+
+  handleChangePassword(e) {
+    // console.log('handle password', e);
```
Remove `console.log`

paolina92 commented on assembl/static2/js/app/pages/changePassword.jsx:
```diff
@@ -1,22 +1,90 @@
+      if (slug) {
+        const route = `/${Routes.get('home', { slug: slug })}`;
+        const url = new URL(route, window.location.href);
+        // console.log(url);
```
Remove `console.log`

paolina92 commented on assembl/static2/js/app/pages/changePassword.jsx:
```diff
@@ -1,22 +1,90 @@
+    .catch((error) => {
+      try {
+        const resp = JSON.parse(error);
+        // TODO: Use the new Alert system from Phase1 branch
```
In this case, the TODO is ok because we have to wait for my branch.

paolina92 commented on assembl/static2/js/app/pages/requestPasswordChange.jsx:
```diff
@@ -0,0 +1,36 @@
+
+class RequestPasswordChange extends React.Component {
+  render() {
+    // console.log('props', this.props);
```
Remove

paolina92 commented on assembl/static2/js/app/pages/requestPasswordChange.jsx:
```diff
@@ -0,0 +1,36 @@
+        <Row className="max-container center">
+          <Col xs={12} md={6} className="col-centered">
+            {!pwdSendConfirm && <SendPwdForm />}
+            {pwdSendConfirm && <SendPwdConfirm />}
```
`{pwdSendConfirm ? <SendPwdConfirm /> : <SendPwdForm />}`

paolina92 commented on assembl/static2/js/app/routes.jsx:
```diff
@@ -14,27 +19,120 @@ import Profile from './pages/profile';
+  eg. router.getRouteFor("login", {slug: 'discussion_slug'})
+  // => '/discussion_slug/login'
+*/
+
```
Same comment than before...

swergas commented on assembl/static2/js/app/services/authenticationService.js:
```diff
@@ -0,0 +1,49 @@
+
+export const signUp = (payload) => {
+  if (payload.password1 !== payload.password2) {
+    alert("Passwords don't match");
```
I guess this needs i18n

paolina92 commented on assembl/static2/js/app/routes.jsx:
```diff
@@ -14,27 +19,120 @@ import Profile from './pages/profile';
+  // => '/discussion_slug/login'
+*/
+
+class RoutesMap {
```
`RoutesMap` should not be here but in a specific module in `utils` for example.

ayyazdaniaryan replied:
Okay, will move

paolina92 commented on assembl/static2/js/app/routes.jsx:
```diff
@@ -14,27 +19,120 @@ import Profile from './pages/profile';
+    this.routes = {
+      oldLogout: 'legacy/logout',
+      oldLogin: 'debate/login',
+      oldDebate: 'debate/${slug}',
```
This class is very useful but for me, it's a problem if `${slug}` is undefined here. We should try another way to get the slug, maybe in the html template? Then we can create a method to get it like `discussionId` for example... Can we discuss about it together before the merge?

ayyazdaniaryan replied:
I think after several discussions, this is more clear to you now, yeah?

swergas commented on assembl/static2/js/app/utils/inputHandler.js:
```diff
@@ -0,0 +1,13 @@
+  @params inputEvent [SyntheticEvent] The event that is triggered from the onChange
+  @returns void 
+*/
+export default (component, inputEvent) => {
```
Is this the right way to bind an input to a state property? Can't we set an attribute on the React element (HTML tag) for this, like in Angular data-model something?

swergas commented on assembl/static2/js/app/utils/inputHandler.js:
```diff
@@ -0,0 +1,12 @@
+  @params inputEvent [SyntheticEvent] The event that is triggered from the onChange
+  @returns void
+*/
+export default (component, inputEvent) => {
```
Is this the right way to bind an input to a state property? Can't we set an attribute on the React element (HTML tag) for this, like in Angular data-model something?

ayyazdaniaryan replied:
Yes, you add an onChange handler for each attribute. Then you have to create many methods to manage the onChange, or create one, and then you have duplication of code. So, I abstracted the duplicated piece of code and put it here. It's to simplify the process, and it is not a globally unique thing. It's only used for form inputs. I'll add comments to be more clear.

maparent commented on assembl/views/__init__.py:
```diff
@@ -101,33 +104,87 @@ def get_theme_info(discussion):
     react_url = '/static2'
     use_webpack_server = asbool(config.get('use_webpack_server'))
     if use_webpack_server:
+        webpack_host = config.get(
```
Done.

maparent commented on assembl/views/auth/views.py:
```diff
@@ -806,15 +804,17 @@ def request_password_change(request):
-            identifier=identifier,
-            slug_prefix="/" + slug if slug else "",
-            title=localizer.translate(_('I forgot my password')))
+        # This is not how to do this!
```
OK, that was unhelpful, apologies. We were working closely together and adding notes for future work rather than thinking how it could be received.
What I saw you were trying to do is to check the provenance (backbone/react main view), and change the login view accordingly. I do not believe that auth view technology should be determined by main view technology, but we should have a switch to use one or the other auth view globally, independent of which main view we come from.
We did not implement a switch but changed everything globally, btw. Not the cleanest, but we were working under absurd time constraints. In my view, we did 85% of the work, and many of your valid comments reflect that.

maparent commented on assembl/views/api2/auth.py:
```diff
@@ -287,6 +301,186 @@ def verify_password(request):
+    context=ClassContext, ctx_class=AgentProfile, header=FORM_HEADER,
+    request_method='POST', permission=NO_PERMISSION_REQUIRED,
+    name="password_reset")
+def reset_password(request):
```
Also see my response to the next comment. Basically, we tried to do a gradual shift, where both mechanisms could coexist, and then shifted to doing a more brutal shift, but did not yet finish killing old code. So yes, you are correct, there is duplication. I do not believe this PR to be complete in that way. As it is, it does add technical debt.

maparent commented on assembl/views/api/token.py:
```diff
@@ -36,3 +40,13 @@ def auth_token(request, extra_headers=None):
+
+@view_config(route_name='check_password_token', request_method='GET',
+             renderer="json")
+def check_password_token(request):
```
Correct. This is something we thought we might use, didn't, yet think we may still use someday. It allows the frontend to know that it is allowed to load on a closed discussion even if the user is not yet in the session.

ayyazdaniaryan commented on assembl/static2/css/components/login.scss:
```diff
@@ -43,6 +43,11 @@
+      color: red;
+      font-weight: bold;
+    }
   }
```
Yep! This is a temporary measure until I use your alert system.

ayyazdaniaryan commented on assembl/static2/js/app/actions/authenticationActions.js:
```diff
@@ -0,0 +1,63 @@
+    }).catch((error) => {
+      // Should match on a more generic thing that accounts for JS errors +
+      // server returned errors
+      switch (error.name) {
```
I think I understand what you mean. I have, for example, created an exception class for mismatched passwords. Do you mean we should seek to create more exception classes that show specific conditions and match on those classes?

ayyazdaniaryan commented on assembl/static2/js/app/actions/contextActions.js:
```diff
@@ -1,14 +1,15 @@
-    dispatch(resolvedAddContext(rootPath, debateId, connectedUserId));
+    dispatch(resolvedAddContext(rootPath, debateId, connectedUserId, connectedUserName));
   };
 };
```
👍

ayyazdaniaryan commented on assembl/static2/js/app/components/common/navbar.jsx:
```diff
@@ -6,6 +6,7 @@ import Avatar from './avatar';
 import LanguageMenu from './languageMenu';
 import NavigationMenu from './navigationMenu';
 import Search from '../search';
+import { Routes } from '../../routes';
```
I know we have discussed moving Routes out of the routes.jsx, and into another file, but can I ask what the issue is here?

ayyazdaniaryan commented on assembl/static2/js/app/components/common/navigationMenu.jsx:
```diff
@@ -2,17 +2,18 @@ import React from 'react';
   render() {
     const { debateData } = this.props.debate;
-    const { rootPath } = this.props.context;
+    // TODO: When turnling url into Link, do not forget activeClassName="active"
```
Ahh, no no, this comment is for if we ever make a `<Link>` further below. There was an activeClassName on there. This comment is not very relevant at the moment. I should re-phrase it.

ayyazdaniaryan commented on assembl/static2/js/app/components/login/assemblLogin.jsx:
```diff
@@ -3,38 +3,42 @@ import { connect } from 'react-redux';
   render() {
-    const { debateData } = this.props.debate;
-    const { rootPath } = this.props.context;
+    const slug = this.props.slug;
```
👍

ayyazdaniaryan commented on assembl/static2/css/errorPages.css:
```diff
@@ -0,0 +1,217 @@
+	bottom: 0;
+	color: #a0a0a0;
+	font-size: 14px
+}
```
This file is completely outside of the V2 and V1 world. It is used for all jinja2 templates that will replace when we get 404s, 500s, etc errors from Pyramid.

ayyazdaniaryan commented on doc/testing.rst:
```diff
@@ -257,6 +257,18 @@ Assembl has several core fixtures that are important to note, in order to run th
+
+
+Getting Started
+^^^^^^^^^^^^^^^
```
Accidentally added during a rebase. Will remove.

ayyazdaniaryan commented on assembl/tests/views_tests/test_browser.py:
```diff
@@ -13,15 +13,19 @@ def test_mocha(browser, test_server, discussion, test_session,
     assert not extractor.failures_number
 
+
+@pytest.mark.xfail
```
I'm actually not sure. @maparent ?

ayyazdaniaryan commented:
Sweet jesus, so many changes since the last PR comment. Major changes:
* `App` is no longer the root component of frontend, it is now `Root`
* `Alert` component is alive everywhere in the application.
* Tons of CSS fixes in the signup page
* Webpack configs updated to support latest version of webpack
* Moved the `RouteMap` to it's own module
* Adapted fixes from merging `phase1_ui` items
* Misc changes (linter fixes, typos, more comments, etc)
